types/felt.go

```diff
@@ -7,106 +7,147 @@ import (
 	"fmt"
 	"io"
 	"math/big"
+	"regexp"
 	"strconv"
 	"strings"
 )
 
+const (
+	FIELD_GEN   int    = 3
+	FIELD_PRIME string = "3618502788666131213697322783095070105623107215331596699973092056135872020481"
+)
+
+var (
+	MaxFelt   = StrToFelt(FIELD_PRIME)
+	utfRegexp = regexp.MustCompile(`\w+`)
+)
+
+// Felt represents Field Element or Felt from cairo.
 type Felt struct {
 	*big.Int
 }
 
-func (f *Felt) Scan(src interface{}) error {
-	var i sql.NullString
-	if err := i.Scan(src); err != nil {
-		return err
-	}
-	if !i.Valid {
-		return nil
+// Big converts a Felt to its big.Int representation.
+func (f *Felt) Big() *big.Int {
+	return new(big.Int).SetBytes(f.Int.Bytes())
+}
+
+// StrToFelt converts a string containing a decimal, hexadecimal or UTF8 charset into a Felt.
+func StrToFelt(str string) *Felt {
+	f := new(Felt)
+	if ok := f.strToFelt(str); ok {
+		return f
 	}
-	if f.Int == nil {
-		f.Int = big.NewInt(0)
+	return nil
+}
+
+func (f *Felt) strToFelt(str string) bool {
+	if b, ok := new(big.Int).SetString(str, 0); ok {
+		f.Int = b
+		return ok
 	}
-	// Value came in a floating point format.
-	if strings.ContainsAny(i.String, ".+e") {
-		flt := big.NewFloat(0)
-		if _, err := fmt.Sscan(i.String, f); err != nil {
-			return err
+	if utfRegexp.MatchString(str) {
+		hexStr := hex.EncodeToString([]byte(str))
+		if b, ok := new(big.Int).SetString(hexStr, 16); ok {
+			f.Int = b
+			return ok
```

Comment:
Even if I think that is a very good idea. I can think of a case where it will not work and that is when we have an array of shorts (i.e. len/Felt*) to represent a longer string and it as a member that is made of number only. There is such a case in the way aspect.co has built their NFT tokenURI. I think we may want to have a specific set of methods (or a  parameter) to manage short strings.

Reply:
I’m not opposed to a specific short string methods. Can you point me to the aspect tokenURI code. I’m not sure I understand why this in conjunction with your ascii regex change wouldn’t do the trick over each array element.

Reply:
This is the line on the aspectco repository https://github.com/aspectco/cairo-contracts/blob/48638b84fb79b02734e0e47d2ef230d57399a212/openzeppelin/utils/ShortString.cairo#L89

The idea is that they return an array of Felts that represents the tokenURI for the NFT. The last 2 Felts are actually a low/high or an uint256 transformed in short-string, thus the `uint256_to_ss` function. TL;DR: the 2 Felts only contain numbers, yet they are short strings and should be translated as such and not as *big.Int

Reply:
So in this case `ToFelt` would just treat the low and high like a big.Int and parse that representation which would be expected, correct?

I could add another function for `ToShortString` giving the string representation, but I'm not sure I understand why the `ToFelt` wouldn't act as expected in this case.